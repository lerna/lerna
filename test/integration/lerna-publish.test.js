"use strict";

const execa = require("execa");
const fs = require("fs-extra");
const globby = require("globby");
const normalizeNewline = require("normalize-newline");
const writeJsonFile = require("write-json-file");
const loadJsonFile = require("load-json-file");
const tempWrite = require("temp-write");
const path = require("path");
const os = require("os");

const cliRunner = require("../helpers/cli-runner");
const initFixture = require("../helpers/initFixture");
const loadPkgManifests = require("../helpers/loadPkgManifests");
const normalizeTestRoot = require("../helpers/normalize-test-root");

// stabilize changelog commit SHA and datestamp
expect.addSnapshotSerializer({
  print(val) {
    return normalizeNewline(val)
      .replace(/\b[a-f0-9]{7,8}\b/g, "SHA")
      .replace(/\(\d{4}-\d{2}-\d{2}\)/g, "(YYYY-MM-DD)");
  },
  test(val) {
    return val && typeof val === "string";
  },
});

const lastCommitMessage = cwd =>
  execa.stdout("git", ["log", "-1", "--format=%B"], { cwd }).then(normalizeNewline);

async function commitChangeToPackage(cwd, packageName, commitMsg, data) {
  const packageJSONPath = path.join(cwd, "packages", packageName, "package.json");
  const pkg = await loadJsonFile(packageJSONPath);

  await writeJsonFile(packageJSONPath, Object.assign(pkg, data));
  await execa("git", ["add", "."], { cwd });

  let gitCommit;

  if (commitMsg.indexOf(os.EOL) > -1) {
    // Use tempfile to allow multi\nline strings.
    const tmpFilePath = await tempWrite(commitMsg);

    gitCommit = execa("git", ["commit", "-F", tmpFilePath], { cwd });
  } else {
    gitCommit = execa("git", ["commit", "-m", commitMsg], { cwd });
  }

  return gitCommit;
}

describe("lerna publish", () => {
  const currentDirectory = process.cwd();

  afterEach(() => {
    // conventional-recommended-bump is incapable of accepting cwd config :P
    if (process.cwd() !== currentDirectory) {
      process.chdir(currentDirectory);
    }
  });

  test("exit 0 when no updates", async () => {
    const cwd = await initFixture("PublishCommand/normal");
    const args = ["publish"];

    await execa("git", ["tag", "-a", "v1.0.0", "-m", "v1.0.0"], { cwd });

    const { code, stdout } = await cliRunner(cwd)(...args);

    expect(code).toBe(0);
    expect(stdout).toBe("");
  });

  test("updates fixed versions", async () => {
    const cwd = await initFixture("PublishCommand/normal");
    const args = ["publish", "--skip-npm", "--cd-version=patch", "--yes"];

    const { stdout } = await cliRunner(cwd)(...args);
    expect(stdout).toMatchSnapshot("stdout");

    const [allPackageJsons, commitMessage] = await Promise.all([
      loadPkgManifests(cwd),
      lastCommitMessage(cwd),
    ]);

    expect(allPackageJsons).toMatchSnapshot("packages");
    expect(commitMessage).toMatchSnapshot("commit");
  });

  test("updates all transitive dependents", async () => {
    const cwd = await initFixture("PublishCommand/snake-graph");
    const args = ["publish", "--skip-npm", "--cd-version=major", "--yes"];

    await execa("git", ["tag", "v1.0.0", "-m", "v1.0.0"], { cwd });
    await commitChangeToPackage(cwd, "package-1", "change", { change: true });

    await cliRunner(cwd)(...args);

    expect(await loadPkgManifests(cwd)).toMatchSnapshot();
  });

  test("uses default suffix with canary flag", async () => {
    const cwd = await initFixture("PublishCommand/normal");
    const args = ["publish", "--canary", "--skip-npm", "--yes"];

    const { stdout } = await cliRunner(cwd)(...args);
    expect(stdout).toMatchSnapshot("stdout");
  });

  test("updates independent versions", async () => {
    const cwd = await initFixture("PublishCommand/independent");
    const args = ["publish", "--skip-npm", "--cd-version=major", "--yes"];

    const { stdout } = await cliRunner(cwd)(...args);
    expect(stdout).toMatchSnapshot("stdout");

    const [allPackageJsons, commitMessage] = await Promise.all([
      loadPkgManifests(cwd),
      lastCommitMessage(cwd),
    ]);

    expect(allPackageJsons).toMatchSnapshot("packages");
    expect(commitMessage).toMatchSnapshot("commit");
  });

  ["normal", "independent"].forEach(flavor =>
    test(`${flavor} mode --conventional-commits changelog`, async () => {
      const cwd = await initFixture(`PublishCommand/${flavor}`, "feat: init repo");
      const args = ["publish", "--conventional-commits", "--skip-git", "--skip-npm", "--yes"];

      await commitChangeToPackage(cwd, "package-1", "feat(package-1): Add foo", { foo: true });
      await commitChangeToPackage(cwd, "package-1", "fix(package-1): Fix foo", { foo: false });
      await commitChangeToPackage(cwd, "package-2", "fix(package-2): Fix bar", { bar: true });
      await commitChangeToPackage(
        cwd,
        "package-3",
        `feat(package-3): Add baz feature${os.EOL}${os.EOL}BREAKING CHANGE: yup`,
        { baz: true }
      );

      // conventional-recommended-bump is incapable of accepting cwd config :P
      process.chdir(cwd);

      const { stdout } = await cliRunner(cwd)(...args);
      expect(stdout).toMatchSnapshot();

      const changelogFilePaths = await globby(["**/CHANGELOG.md"], {
        cwd,
        absolute: true,
        followSymlinkedDirectories: false,
      });
      const changelogContents = await Promise.all(
        changelogFilePaths.sort().map(fp => fs.readFile(fp, "utf8"))
      );

      expect(changelogContents).toMatchSnapshot();
    })
  );

  it("replaces file: specifier with local version before npm publish but after git commit", async () => {
    const cwd = await initFixture("PublishCommand/relative-file-specs");

    await execa("git", ["tag", "v1.0.0", "-m", "v1.0.0"], { cwd });
    await commitChangeToPackage(cwd, "package-1", "feat(package-1): Add foo", { foo: true });

    await cliRunner(cwd)("publish", "--cd-version=major", "--skip-npm", "--yes");

    expect(
      await execa.stdout("git", ["show", "--unified=0", "--ignore-space-at-eol", "--format=%s"], { cwd })
    ).toMatchSnapshot("committed");

    expect(
      await execa.stdout("git", ["diff", "--unified=0", "--ignore-space-at-eol"], { cwd })
    ).toMatchSnapshot("unstaged");
  });

  test("calls lifecycle scripts", async () => {
    const cwd = await initFixture("PublishCommand/lifecycle");
    const args = ["publish", "--skip-npm", "--cd-version", "minor", "--yes"];

    const { stdout } = await cliRunner(cwd)(...args);
    expect(normalizeTestRoot(cwd)(stdout)).toMatchSnapshot();
  });

  test("silences lifecycle scripts with --loglevel=silent", async () => {
    const cwd = await initFixture("PublishCommand/lifecycle");
    const args = ["publish", "--skip-npm", "--cd-version", "minor", "--yes", "--loglevel", "silent"];

    const { stdout } = await cliRunner(cwd)(...args);
    expect(normalizeTestRoot(cwd)(stdout)).toMatchSnapshot();
  });
});
