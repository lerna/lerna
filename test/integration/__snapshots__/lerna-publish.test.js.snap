// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commit: updates fixed versions 1`] = `
"v1.0.1
"
`;

exports[`commit: updates independent versions 1`] = `
"Publish

 - package-1@2.0.0
 - package-2@3.0.0
 - package-3@4.0.0
 - package-4@5.0.0
 - package-5@6.0.0
"
`;

exports[`packages: updates fixed versions 1`] = `
Array [
  Object {
    "name": "package-1",
    "version": "1.0.1",
  },
  Object {
    "dependencies": Object {
      "package-1": "^1.0.1",
    },
    "name": "package-2",
    "version": "1.0.1",
  },
  Object {
    "devDependencies": Object {
      "package-2": "^1.0.1",
    },
    "name": "package-3",
    "version": "1.0.1",
  },
  Object {
    "dependencies": Object {
      "package-1": "^0.0.0",
    },
    "name": "package-4",
    "version": "1.0.1",
  },
  Object {
    "dependencies": Object {
      "package-1": "^1.0.1",
    },
    "name": "package-5",
    "private": true,
    "version": "1.0.1",
  },
]
`;

exports[`packages: updates independent versions 1`] = `
Array [
  Object {
    "name": "package-1",
    "version": "2.0.0",
  },
  Object {
    "dependencies": Object {
      "package-1": "^2.0.0",
    },
    "name": "package-2",
    "version": "3.0.0",
  },
  Object {
    "devDependencies": Object {
      "package-2": "^3.0.0",
    },
    "name": "package-3",
    "version": "4.0.0",
  },
  Object {
    "dependencies": Object {
      "package-1": "^0.0.0",
    },
    "name": "package-4",
    "version": "5.0.0",
  },
  Object {
    "dependencies": Object {
      "package-3": "^4.0.0",
    },
    "name": "package-5",
    "private": true,
    "version": "6.0.0",
  },
]
`;

exports[`stderr: exists with a exit code 0 when there are no updates 1`] = `
"lerna info version __TEST_VERSION__
lerna info current version 1.0.0
lerna info Checking for updated packages...
lerna info Comparing with tag v1.0.0
lerna info No updated packages to publish. "
`;

exports[`stderr: updates fixed versions 1`] = `
"lerna info version __TEST_VERSION__
lerna info current version 1.0.0
lerna info Checking for updated packages...
lerna WARN No tags found!
lerna info Comparing with initial commit.
lerna info auto-confirmed "
`;

exports[`stderr: updates independent versions 1`] = `
"lerna info version __TEST_VERSION__
lerna info versioning independent
lerna info Checking for updated packages...
lerna WARN No tags found!
lerna info Comparing with initial commit.
lerna info auto-confirmed "
`;

exports[`stdout: exists with a exit code 0 when there are no updates 1`] = `""`;

exports[`stdout: updates fixed versions 1`] = `
"
Changes:
 - package-1: 1.0.0 => 1.0.1
 - package-2: 1.0.0 => 1.0.1
 - package-3: 1.0.0 => 1.0.1
 - package-4: 1.0.0 => 1.0.1
 - package-5: 1.0.0 => 1.0.1 (private)
"
`;

exports[`stdout: updates independent versions 1`] = `
"
Changes:
 - package-1: 1.0.0 => 2.0.0
 - package-2: 2.0.0 => 3.0.0
 - package-3: 3.0.0 => 4.0.0
 - package-4: 4.0.0 => 5.0.0
 - package-5: 5.0.0 => 6.0.0 (private)
"
`;
